src/java/org/apache/cassandra/db/partitions/BTreePartitionUpdate.java

```diff
@@ -0,0 +1,685 @@
+    }
+
+    /**
+     * Turns the given iterator into an update.
+     *
+     * @param iterator the iterator to turn into updates.
+     *
+     * Warning: this method does not close the provided iterator, it is up to
+     * the caller to close it.
+     */
+    @SuppressWarnings("resource")
+    public static BTreePartitionUpdate fromIterator(UnfilteredRowIterator iterator)
+    {
+        BTreePartitionData holder = build(iterator, 16);
+        MutableDeletionInfo deletionInfo = (MutableDeletionInfo) holder.deletionInfo;
+        return new BTreePartitionUpdate(iterator.metadata(), iterator.partitionKey(), holder, deletionInfo, false);
+    }
+
+    /**
+     * Turns the given iterator into an update.
+     *
+     * @param iterator the iterator to turn into updates.
+     * @param filter the column filter used when querying {@code iterator}. This is used to make
+     * sure we don't include data for which the value has been skipped while reading (as we would
+     * then be writing something incorrect).
+     *
+     * Warning: this method does not close the provided iterator, it is up to
+     * the caller to close it.
+     */
+    @SuppressWarnings("resource")
+    public static BTreePartitionUpdate fromIterator(UnfilteredRowIterator iterator, ColumnFilter filter)
+    {
+        return fromIterator(UnfilteredRowIterators.withOnlyQueriedData(iterator, filter));
+    }
+
+    protected boolean canHaveShadowedData()
+    {
+        return canHaveShadowedData;
+    }
+
+    /**
+     * Creates a partition update that entirely deletes a given partition.
+     *
+     * @param metadata the metadata for the created update.
+     * @param key the partition key for the partition that the created update should delete.
+     * @param timestamp the timestamp for the deletion.
+     * @param nowInSec the current time in seconds to use as local deletion time for the partition deletion.
+     *
+     * @return the newly created partition deletion update.
+     */
+    public static BTreePartitionUpdate fullPartitionDelete(TableMetadata metadata, ByteBuffer key, long timestamp, int nowInSec)
+    {
+        return fullPartitionDelete(metadata, metadata.partitioner.decorateKey(key), timestamp, nowInSec);
+    }
+
+    public static BTreePartitionUpdate asBTreeUpdate(PartitionUpdate update)
+    {
+        if (update instanceof BTreePartitionUpdate)
+            return (BTreePartitionUpdate) update;
+
+        try (UnfilteredRowIterator iterator = update.unfilteredIterator())
+        {
+            return fromIterator(iterator);
+        }
+    }
+
+    /**
+     * Merges the provided updates, yielding a new update that incorporates all those updates.
+     *
+     * @param updates the collection of updates to merge. This shouldn't be empty.
+     *
+     * @return a partition update that include (merge) all the updates from {@code updates}.
+     */
+    public PartitionUpdate merge(List<? extends PartitionUpdate> updates)
```

Comment:
I find this switch from static to non-static method confusing. When I read that initially, I fully expected that the method was changed to merge `this` with the remaining `updates`, and so was confused by the implementation (and initially though this was an oversight). I understand now `this` is only used as a way to pick the underlying implementation class, but this seems unexpected and thus error prone to me. I'd suggest any one of the following alternatives:
1. keep the method here, but make it be `this` merged to `updates` (that wouldn't include `this`). Not my preferred option, but I'd be ok with it.
2. if we want to keep it essentially "static" (not truly applying to `this` directly) but depending on the underlying implementation, then I think this should move as a `Factory` method (the factory is in my mind just a way to get "static yet implementation dependent" methods). And add a `factory()` method to `PartitionUpdate` (that obviously returns the factory for the concrete object you call it on).
3. but since ultimately this does not currently use any implementation specific optimisations, this could also be a static method of `PartitionUpdate` directly if we prefer.

Fwiw, my preference would be to do option 2 above, but as a default method of `PartitionUpdate.Factory` to avoid code duplication.

Reply:
Added a factory method as well as static `PartitionUpdate` one that calls it, using the factory of the first update in the list.